DynamicMocks: add onMockFor to provide for dynamic mock injection

haikusw opened the pull request:
Sometimes you want the same URL to return different results at different times during a test (for example, an error payload first and then success payload).  Dynamic Mock support enables this.

The particular use case I had was for a service call made with an expired token. The networking layer in this app will catch the expired token error, automatically make a `refresh token` service call, and then auto-retry the original service call.

In this situation I needed to have the mock for the original service call return a 401 for the expired token as long as the `currentToken` doesn't match the `newGoodToken` returned by the mocked refresh token call (which is auto-saved into `currentToken` by the networking layer).

So my (simplified) `onMockFor` will look something like this:

```

let refreshMock = Mock(url: refreshURL, dataType: .json, statusCode: 200, data: [.get: goodRefreshTokenData])
refreshMock.register()
let failureMock = Mock(url: serviceURL, dataType: .json, statusCode: 401, data: [.get, Data()])
failureMock.register()

let successMock = Mock(url: serviceURL, dataType: .json, statusCode: 200, data: [.get: goodMockData])
// Note: do not register! Will overwrite failureMock (same URL).

// Then add the `onMockFor` closure to check if we have the new token
Mocker.onMockFor = { (request: URLRequest) -> Mock? in
  guard request.url == serviceURL else { return nil }
  if currentToken == newGoodToken {
      return successMock
  }
  return nil
}
```

haikusw committed:
DynamicMocks: add onMockFor to provide for dynamic mock injection

Sometimes you want a url to return different things for the same URL during a test (error and then success, for example).  Dynamic Mock support enables that.

haikusw committed:
DynamicMocks - make onMockFor accessible without @testable import

haikusw committed:
Dynamic Mocks: fix indentation to match project settings.

(tabs to spaces)

woolie commented on MockerTests/MockerTests.swift:
```diff
@@ -401,4 +408,126 @@ final class MockerTests: XCTestCase {
+        wait(for: [expectation], timeout: 2.0)
+
+		XCTAssert(onMockForCalled, "our onMockFor handler was not called!")
+	}
```
Mix of TAB and spaces for indentation makes this piece of code look strangely indented,

woolie commented on Sources/Mocker.swift:
```diff
@@ -75,11 +80,20 @@ public struct Mocker {
     /// - Returns: A mock if found, `nil` if there's no mocked data registered for the given request.
     static func mock(for request: URLRequest) -> Mock? {
         shared.queue.sync {
+			/// Zeroth support for dynamic mocks via `onMockFor` closure
```
Was the use of Zeroth on purpose or a typo?

haikusw replied:
on purpose, but misguided 2am decision :) Fixed. Thanks.

woolie commented:
Just one question about a comment. Otherwise looks fine to me.

haikusw committed:
Dynamic Mocks: improve comment

/// comments internal to a function do not show up in Quick Help so move that information to the function comment and demote internal comments to //

haikusw committed:
Update MockerTests/MockerTests.swift

Co-authored-by: Boris Emorine <boris.emorine@gmail.com>

haikusw committed:
Update MockerTests/MockerTests.swift

Co-authored-by: Boris Emorine <boris.emorine@gmail.com>

haikusw committed:
Update Sources/Mocker.swift

Co-authored-by: Boris Emorine <boris.emorine@gmail.com>

haikusw committed:
Update MockerTests/MockerTests.swift

Co-authored-by: Boris Emorine <boris.emorine@gmail.com>

haikusw committed:
Dynamic Mock: Update Readme as suggested

haikusw committed:
Merge branch 'dynamic-mock' of https://github.com/haikusw/Mocker into dynamic-mock

haikusw committed:
Make an example test and service to test to show the reason for onMockFor

This commit contains a test file with a simplified example of a service one wants to test and shows how onMockFor is used to test it.

I was not able to use existing Mocker functionality to create the test to test this functionality.  Might be a way, but I’m not seeing it. Hence onMockFor which, as you can see in this test, works perfectly.

haikusw committed:
convert tabs to spaces

Sorry - current client project uses tabs so that’s Xcode’s default on my machine currently.

haikusw commented:
Hey woolie, can you let me know if you got this message?  Testing of comments on closed PRs get sent. Thx.

haikusw committed:
Add my understanding of the two Tests AvdLee suggested

Might be mis-understanding so it seemed best to write my understanding in code to aid communication.

haikusw committed:
Set project settings to spaces to match project convention

Changed the project setting to spaces to match the convention for the project so that I don’t keep accidentally submitting tabs (my Xcode default seting because of current client requirements).  This will help other contributors avoid the same issue.
